### ISSUE TITLE
Issue with Kryo5Codec in combination with org.springframework.cache.support.NullValue

### ISSUE THREAD

**Issue by MrKanister2000** [redisson/redisson#5830]
Hi,

I'm currently having an issue with the Spring Cache implementation in combination with Redisson. 
The JCacheCache class provided by Redisson extends `org.springframework.cache.support.AbstractValueAdaptingCache`. In the method `protected Object fromStoreValue(@Nullable Object storeValue)` of the class AbstractValueAdaptingCache, there is an if condition, that check if the storeValue is equal to NullValue.INSTANCE using ==. 

```
protected Object fromStoreValue(@Nullable Object storeValue) {
	if (this.allowNullValues && storeValue == NullValue.INSTANCE) {
		return null;
	}
	return storeValue;
}
```

This condition evaluated to false in my case, because the instance of storeValue was not the same instance as NullValue.INSTANCE. Reason is the deserialisation, that was done by Kryo. It seems, that Kryo changes the constructor to "public" using reflections and creates a new instance by calling the constructor. It does not call the "readResolve()" method of NullValue class, which would return NullValue.INSTANCE.
Is this a known issue? The only solution I came up with is extending the Kryo5Codec and adding a custom Serializer for NullValue.class. Is there another way to fix this issue?

Best regards

**Comment by mrniko** [comment]
Can you add the code below into org.redisson.codec.Kryo5Codec#createKryo method and say if it works?

```java
if (com.esotericsoftware.kryo.util.Util.isClassAvailable("org.springframework.cache.support.NullValue")) {
   kryo.addDefaultSerializer(Class.forName("org.springframework.cache.support.NullValue"), new JavaSerializer());
}

**Comment by MrKanister2000** [comment]
Yes, it works, thanks. Do you see any trade-offs (like performance) when using the JavaSerializer?

My first solution was extending the Kryo5Codec class:

```
public class MyKryo5Codec extends Kryo5Codec {

    @Override
    protected Kryo createKryo(ClassLoader classLoader) {
        Kryo kryo = super.createKryo(classLoader);

        kryo.addDefaultSerializer(NullValue.class, new NullValueSerializer(kryo, NullValue.class));

        return kryo;
    }
}
```

and creating a custom NullValueSerializer:

```
public class NullValueSerializer extends FieldSerializer<NullValue> {

    public NullValueSerializer(Kryo kryo, Class type) {
        super(kryo, type);
    }

    @Override
    public NullValue read(Kryo kryo, Input input, Class type) {
        return (NullValue) NullValue.INSTANCE;
    }
}
```

But your solution has a way smaller footprint than mine.

**Comment by mrniko** [comment]
Thanks for testing. In your example Spring become a required dependency because of explicit NullValue class definition which I would like to avoid.
